### ISSUE TITLE
✏️ [ FEAT REQ ] Encrypted text messages with other Quik users

### ISSUE THREAD

**Issue by Dyras** [octoshrimpy/quik#55]
**Is your feature request related to a problem? If so, please describe the problem.**
There is damn near no security for regular SMS at all. Sure, RCS is slowly coming, but that seems to be Google Messages exclusive + it's not *really* SMS. You're still using data for it.

Several apps have tried it, but they are either abandoned (Silence, Partisan SMS) or too new to be used in practice (Deku SMS).

Hence I would suggest..

**Describe the feature you'd like**
The ability to encrypt your SMS messages with other Quik users. I think that so far, Silence has done it the best with forward secrecy, etc.

Basically, I want the ability to opt-in to using encryption with other people using Quik, so that the telephone company can't read the contents of your text messages.

**Additional context**
Yes, I know that (insert non-SMS app) is encrypted. On the rare occasion that I send an SMS with a friend, I would still prefer it if that was encrypted rather than unencrypted as is the case with darn near every SMS sent worldwide today.

Additionally, in some cases SMS will be the only way of communicating, for example, if the data network isn't working.

**Comment by octoshrimpy** [comment]
I once cooked up an idea of using [stegcloak](https://stegcloak.surge.sh) and the first couple SMS for a handshake on cryptography, but haven't really tried it before.

**Comment by octoshrimpy** [comment]
---

main system


first message sent includes `secure layer` message containing 
> Alice: "hey I'm on matrix, are you?"


second message sent will not include any more `secure layer` content, unless user specifies to resend handshake within conversation settings
> Alice: (N/A)


upon receiving a `secure layer` request, the client will prompt the user for handshake
> Bob - [ Alice wants to handshake ]


upon handshake request, user may decide one of three things
1. > Bob - yes
  * prompts end of handshake
  * convo will be moved to new protocol

2. > Bob - no
  * prompts end of handshake
  * convo will remain where it is
  * both clients will no longer attempt to find `secure layer` messages

3. > Bob - "ghost"
  * does not end handshake
  * Bob's client pretends it never saw the message and is on a non-moxie client
  * Bob's client will no longer attempt to find `secure layer` messages


> Bob - no
upon receiving no, client will no longer attempt to find `secure layer` messages


> Bob - "ghost"
client will never receive "ghost", as it's never sent


> Bob - yes
upon receiving handshake "yes", client will prompt user to send own matrix username
> Alice: "alice@home.server"

1. upon receiving hash and username, client can decide whether to finish handshake, or cancel
> Bob - alice@home.server


> Bob - yes
  * finalizes handshake
  * client sends hash of own username as activityPub message
  * client takes received matrix username and inits connection through matrix protocol

> Bob - no
  * prompts end of handshake
  * convo will remain where it is
  * both clients will no longer attempt to find `secure layer` messages

> Bob - "ghost"
  * Bob's client pretends it never saw the message and is on a non-moxie client
  * Bob's client will no longer attempt to find `secure layer` messages

upon receiving a matrix message, check if incoming username hashes to a received handshake
if yes, tie the two together
> Alice 

---


here's the main idea. it was for shifting over to matrix, but the same idea could be used for instead accepting crypto keys

**Comment by Dyras** [comment]
I believe in you man. Make us proud!

Say the word and you have a donation every month to motivate you to figure this out.

**Comment by Dyras** [comment]
If they use the DekuSMS approach then my understanding is that the public keys can be shared over SMS. You're not sending the private keys after all.

**Comment by Anderhar** [comment]
> Several apps have tried it, but they are either abandoned (Silence, Partisan SMS) or too new to be used in practice (Deku SMS)

Nowadays, these two categories can be expanded with [KryptEY](https://github.com/amnesica/KryptEY) (abandoned) and [Vidar](https://github.com/DrSolidDevil/Vidar) (too new). 

This is quite unfortunate, because at a times when more and more governments are tending to restrict their citizens' access to private messaging over the Internet (or to the Internet in general), encrypted SMS becomes almost the only reliable way to communicate safely.
Hope QUIK can assist with that someday as well.

**Comment by eternalevelyn** [comment]
I have a really odd question/suggestion because I'm very very paranoid. I've tried stuff like DresSecureComms and while I like it, I have a very particular preference in cryptography (the "better _way_ safe than sorry" approach). I'm curious- is there any chance y'all would be open to me implementing/maintaining something very custom and novel?

Because I want to implement it with a custom encryption suite for people as paranoid as I am.

Encryption suite would look something like what you've already specified in this thread, but with the following full cryptosystem:

### Post quantum key exchange/signing scheme:
- Key exchange: hybrid ML-KEM-1024/X448, like how gpg does it
- Signing: hybrid ML-DSA-87/Ed448

### Encryption:
- Serpent-256-OCB

### Hash algorithm for your suggested implementation architecture:
- BLAKE3

I'd also be willing to implement/maintain a mode that works via normal (e.g. standard) cryptosystems that y'all agree on additionally, so that people have options between super-paranoid mode and not-quite-so-paranoid mode. You'd be free to remove super-paranoid mode if I implemented both and stopped maintaining the super-paranoid mode.

It's well within your rights to refuse, I just have odd preferences. Let me know what y'all think! Feel free to ask why I have these preferences, I have just a _stupid_ amount of (admittedly paranoid) reasoning to back them up.

**Comment by Anderhar** [comment]
> Several apps have tried it, but they are either abandoned (Silence, Partisan SMS) or too new to be used in practice (Deku SMS).

> Nowadays, these two categories can be expanded with [KryptEY](https://github.com/amnesica/KryptEY) (abandoned) and [Vidar](https://github.com/DrSolidDevil/Vidar) (too new).

Let's add [Kryptos](https://f-droid.org/packages/com.kryptos.android/) to this list. Looks pretty solid at the first glance.

**Comment by a8shjtvcb7m3** [comment]
Any new on this subject or it's dead ? I'd love to have this feature since lot of people I know still only use SMS :(
Side note : someone created a fork of Silence, maybe worth checking out : https://github.com/jimvixx/ShortMessageSecure

**Comment by Inhishonor** [comment]
@a8shjtvcb7m3 Nothing new, but postponed for the time being. Once other stuff are taken care of we can get back to it.
